src/V3Param.cpp

```diff
@@ -52,14 +52,140 @@
+    typedef std::map<string, AstConst*> ParamConstMap;
+    typedef std::map<const V3HierarchicalBlockOption*, ParamConstMap> ParamsMap;
+
+    // key:Original module name, value:HiearchyBlockOption*
+    // If a module is parameterized, the module is uniquiefied to overridden parameters.
+    // This is why HierBlockOptsByOrigName is multimap.
+    HierBlockOptsByOrigName m_hierBlockOptsByOrigName;
+
+    // key:mangled module name, value:AstNodeModule*
+    HierBlockModMap m_hierBlockMod;
+
+    // Overridden parameters of the hierarchical block
+    ParamsMap m_params;
+
+    static bool areSame(AstVar* modvarp, AstConst* varp, AstConst* paramp) {
+        if (varp->isString()) { return varp->num().toString() == paramp->num().toString(); }
+        if (modvarp->declKwd() == AstBasicDTypeKwd::FLOAT
+            || modvarp->declKwd() == AstBasicDTypeKwd::DOUBLE) {
+            return v3EpsilonEqual(varp->num().toDouble(), paramp->num().toDouble());
+        }
+        // Now integer type is assumed
+        // bitwidth of paramp is accurate because V3Width already caluclated in the previous run.
+        // varp is before width analysis.
+        const int bitwidth = paramp->num().width();
+        V3Number varNum(varp, bitwidth);
+        if (varp->isSigned()) {
+            varNum.opExtendS(varp->num(), varp->num().width());
+        } else {
+            varNum.opAssign(varp->num());
+        }
+        V3Number isEq(varp, 1);
+        isEq.opEq(varNum, paramp->num());
+        return isEq.isNeqZero();
+    }
+
+public:
+    ParameterizedHierBlocks(const V3HierBlockOptSet& hierOpts, AstNetlist* nodep) {
+        for (V3HierBlockOptSet::const_iterator it = hierOpts.begin(); it != hierOpts.end(); ++it) {
+            m_hierBlockOptsByOrigName.insert(std::make_pair(it->second.origName(), &it->second));
+            const V3HierarchicalBlockOption::ParamStrMap& params = it->second.params();
+            ParamConstMap& consts = m_params[&it->second];
+            for (V3HierarchicalBlockOption::ParamStrMap::const_iterator pIt = params.begin();
+                 pIt != params.end(); ++pIt) {
+                AstConst* constp = AstConst::parseParamLiteral(
+                    new FileLine(FileLine::EmptySecret()), pIt->second);
+                UASSERT(constp, pIt->second << " is not a valid parameter literal");
+                const bool inserted = consts.insert(std::make_pair(pIt->first, constp)).second;
+                UASSERT(inserted, pIt->first << " is already added");
+            }
+        }
+        for (AstNodeModule* modp = nodep->modulesp(); modp;
+             modp = VN_CAST(modp->nextp(), NodeModule)) {
+            if (hierOpts.find(modp->prettyName()) != hierOpts.end()) {
+                m_hierBlockMod.insert(std::make_pair(modp->name(), modp));
+            }
+        }
+    }
+    ~ParameterizedHierBlocks() {
+        for (ParamsMap::iterator it = m_params.begin(); it != m_params.end(); ++it) {
```

Comment:
const_iterator?

Reply:
You're right.
Done in 820a498347019b5c4d655c2c1b2fc1ef73127084